### ISSUE TITLE
Unity clients fail with IL2CPP builds

### ISSUE THREAD

**Issue by Britvich** [nats-io/nats.net.v1#361]
**Update:** The problem is with Unity. See https://github.com/nats-io/nats.net/issues/361#issuecomment-581632837 (below) for a workaround...

**Update 2:** The workaround doesn't work, at least on iOS devices.

**Update 3:** Got it to work. See https://github.com/nats-io/nats.net/issues/361#issuecomment-593048632 (below)...

**Update 4:** It still fails sometimes even when the `UseOldRequestStyle` option is `true`.

**Update 5:** The problem has been solved. See: https://github.com/nats-io/nats.net/pull/370

I have a Unity project using NATS that works fine if built using Mono on Android or if using IL2CPP development build. If built as a release build with IL2CPP for Android or iOS, the NATS connection disconnects on the second request/reply and attempts to reconnect.

This appears to be similar to issue #358.

Here's the NATS server log...

```
[8125] 2020/01/27 10:19:42.095095 [DBG] 66.60.161.134:40560 - cid:9 - Client connection created
[8125] 2020/01/27 10:19:42.158135 [TRC] 66.60.161.134:40560 - cid:9 - <<- [CONNECT {"verbose":false,"pedantic":false,"user":"","pass":"[REDACTED]","ssl_required":false,"name":"","auth_token":"","lang":".NET","version":"0.0.1","protocol":1,"jwt":"","nkey":"","sig":"","echo":false}]
[8125] 2020/01/27 10:19:42.158195 [TRC] 66.60.161.134:40560 - cid:9 - <<- [PING]
[8125] 2020/01/27 10:19:42.158216 [TRC] 66.60.161.134:40560 - cid:9 - ->> [PONG]
[8125] 2020/01/27 10:19:42.359824 [TRC] 66.60.161.134:40560 - cid:9 - <<- [SUB _INBOX.ec2db33d7faf4825ad7bba87047defaf.*  1]
[8125] 2020/01/27 10:19:42.359986 [TRC] 66.60.161.134:40560 - cid:9 - <<- [PUB Citizen.Login _INBOX.ec2db33d7faf4825ad7bba87047defaf.1 15]
[8125] 2020/01/27 10:19:42.360003 [TRC] 66.60.161.134:40560 - cid:9 - <<- MSG_PAYLOAD: ["\x92\xa4Frog\xa8blobblob"]
[8125] 2020/01/27 10:19:42.360016 [TRC] 127.0.0.1:45964 - cid:1 - ->> [MSG Citizen.Login 2 _INBOX.ec2db33d7faf4825ad7bba87047defaf.1 15]
[8125] 2020/01/27 10:19:42.361726 [TRC] 127.0.0.1:45964 - cid:1 - <<- [PONG]
[8125] 2020/01/27 10:19:42.361753 [TRC] 127.0.0.1:45964 - cid:1 - <<- [PUB _INBOX.ec2db33d7faf4825ad7bba87047defaf.1 11]
[8125] 2020/01/27 10:19:42.361767 [TRC] 127.0.0.1:45964 - cid:1 - <<- MSG_PAYLOAD: ["\x92\x00\xcf\x01\xf7\xf3w\xe6\x00\x00\x00"]
[8125] 2020/01/27 10:19:42.361786 [TRC] 66.60.161.134:40560 - cid:9 - ->> [MSG _INBOX.ec2db33d7faf4825ad7bba87047defaf.1 1 11]
[8125] 2020/01/27 10:19:42.804264 [TRC] 66.60.161.134:40560 - cid:9 - <<- [SUB Avatar.Pose.test  2]
[8125] 2020/01/27 10:19:42.804320 [TRC] 66.60.161.134:40560 - cid:9 - <<- [SUB Avatar.Select.test  3]
[8125] 2020/01/27 10:19:42.804352 [TRC] 66.60.161.134:40560 - cid:9 - <<- [SUB Prop.OnAdd.test  4]
[8125] 2020/01/27 10:19:42.804365 [TRC] 66.60.161.134:40560 - cid:9 - <<- [SUB Prop.OnMove.test  5]
[8125] 2020/01/27 10:19:42.804380 [TRC] 66.60.161.134:40560 - cid:9 - <<- [SUB Prop.OnRemove.test  6]
[8125] 2020/01/27 10:19:42.911325 [TRC] 66.60.161.134:40560 - cid:9 - <<- [PUB Prop.Look.test _INBOX.ec2db33d7faf4825ad7bba87047defaf.2 10]
[8125] 2020/01/27 10:19:42.911356 [TRC] 66.60.161.134:40560 - cid:9 - <<- MSG_PAYLOAD: ["\x91\xcf\x01\xf7\xf3w\xe6\x00\x00\x00"]
[8125] 2020/01/27 10:19:42.911371 [TRC] 127.0.0.1:46110 - cid:2 - ->> [MSG Prop.Look.test 2 _INBOX.ec2db33d7faf4825ad7bba87047defaf.2 10]
[8125] 2020/01/27 10:19:42.914473 [TRC] 127.0.0.1:46110 - cid:2 - <<- [PONG]
[8125] 2020/01/27 10:19:42.914499 [TRC] 127.0.0.1:46110 - cid:2 - <<- [PUB _INBOX.ec2db33d7faf4825ad7bba87047defaf.2 74]
[8125] 2020/01/27 10:19:42.914518 [TRC] 127.0.0.1:46110 - cid:2 - <<- MSG_PAYLOAD: ["\x92\x00\x81\xcf\x01\xeao_\x93\x01\x00\x00\x92\x02\x95\x93\xca\x00\x00\x00\x00\xca\x00\x00\x00\x00\xca\x00\x00\x00\x00\x94\xca\x00\x00\x00\x00\xca\x00\x00\x00\x00\xca\x00\x00\x00\x00\xca?\x80\x00\x00\xacTrackedImage\xa3Bar\xca>L\xcc\xcd"]
[8125] 2020/01/27 10:19:42.914534 [TRC] 66.60.161.134:40560 - cid:9 - ->> [MSG _INBOX.ec2db33d7faf4825ad7bba87047defaf.2 1 74]
[8125] 2020/01/27 10:19:42.967805 [DBG] 66.60.161.134:40560 - cid:9 - Client connection closed
[8125] 2020/01/27 10:19:42.967893 [TRC] 66.60.161.134:40560 - cid:9 - <-> [DELSUB 1]
[8125] 2020/01/27 10:19:42.967910 [TRC] 66.60.161.134:40560 - cid:9 - <-> [DELSUB 2]
[8125] 2020/01/27 10:19:42.967933 [TRC] 66.60.161.134:40560 - cid:9 - <-> [DELSUB 3]
[8125] 2020/01/27 10:19:42.967939 [TRC] 66.60.161.134:40560 - cid:9 - <-> [DELSUB 4]
[8125] 2020/01/27 10:19:42.967944 [TRC] 66.60.161.134:40560 - cid:9 - <-> [DELSUB 5]
[8125] 2020/01/27 10:19:42.967950 [TRC] 66.60.161.134:40560 - cid:9 - <-> [DELSUB 6]
[8125] 2020/01/27 10:19:43.241501 [DBG] 66.60.161.134:40562 - cid:10 - Client connection created
[8125] 2020/01/27 10:19:43.285678 [TRC] 66.60.161.134:40562 - cid:10 - <<- [CONNECT {"verbose":false,"pedantic":false,"user":"","pass":"[REDACTED]","ssl_required":false,"name":"","auth_token":"","lang":".NET","version":"0.0.1","protocol":1,"jwt":"","nkey":"","sig":"","echo":false}]
[8125] 2020/01/27 10:19:43.285773 [TRC] 66.60.161.134:40562 - cid:10 - <<- [PING]
[8125] 2020/01/27 10:19:43.285788 [TRC] 66.60.161.134:40562 - cid:10 - ->> [PONG]
[8125] 2020/01/27 10:19:43.352220 [TRC] 66.60.161.134:40562 - cid:10 - <<- [SUB _INBOX.ec2db33d7faf4825ad7bba87047defaf.*  1]
[8125] 2020/01/27 10:19:43.352270 [TRC] 66.60.161.134:40562 - cid:10 - <<- [SUB Avatar.Pose.test  2]
[8125] 2020/01/27 10:19:43.352300 [TRC] 66.60.161.134:40562 - cid:10 - <<- [SUB Avatar.Select.test  3]
[8125] 2020/01/27 10:19:43.352309 [TRC] 66.60.161.134:40562 - cid:10 - <<- [SUB Prop.OnAdd.test  4]
[8125] 2020/01/27 10:19:43.352374 [TRC] 66.60.161.134:40562 - cid:10 - <<- [SUB Prop.OnMove.test  5]
[8125] 2020/01/27 10:19:43.352406 [TRC] 66.60.161.134:40562 - cid:10 - <<- [SUB Prop.OnRemove.test  6]
[8125] 2020/01/27 10:19:43.444747 [TRC] 66.60.161.134:40562 - cid:10 - <<- [PING]
[8125] 2020/01/27 10:19:43.444806 [TRC] 66.60.161.134:40562 - cid:10 - ->> [PONG]
[8125] 2020/01/27 10:19:45.450564 [DBG] 66.60.161.134:40562 - cid:10 - Client Ping Timer
[8125] 2020/01/27 10:19:45.450651 [TRC] 66.60.161.134:40562 - cid:10 - ->> [PING]
```

**Comment by ColinSullivan1** [comment]
Thanks for the information, this really helps.  Yes, this issue does look related to #358.

Unfortunately this looks to be a unity problem originating way down in a collection class.  That being said, we'll do the best we can to help.

It looks like Unity has had this issue with a number of API calls:
 - https://forum.heroiclabs.com/t/unity-2019-2-background-exception-when-socket-connects/361
 - https://github.com/neuecc/MessagePack-CSharp/issues/748

I searched the unity issue tracker specifically for `Microsoft.Win32.ManifestEtw:EventRegister`, and oddly enough this issue popped up (it was referenced in the other issues as well):
 - https://issuetracker.unity3d.com/issues/unable-to-load-dll-advapi32-dot-dll-when-calling-datetime-dot-now

It looks like unity solved some of these in version 2019.3, and hopefully they addressed this one.  Which version of unity are you using?

**Comment by Britvich** [comment]
Thanks for your help. Here's the environment where it fails:

- Unity 2019.3.0f3 and 2019.3.0f5
Scripting backend: IL2CPP
API Compatibility: .NET Standard 2.0 and .NET 4.x
C++ Compiler Configuration: Release
Target Architectures: ARM64
Allow 'unsafe' Code: On
Strip Engine Code: On
Managed Stripping Level: Low
- MessagePack 1.7.3.7
- NATS.Client 0.10.1-pre2

Here's some shot in the dark things I'm investigating:

- I noticed you recently added:
`t.IsBackground = true;`
Should I try that code or do you think it's unrelated?
- I'm using `low` code stripping, I wonder if it will work with `no` code stripping.
- I noticed you use the `Stopwatch` out of the `System.Diagnostics` namespace. I wonder if that related.
- I wonder if MessagePack `Deserialize` is causing the problem, because that second reply is a lot more complex than the first one as it includes deserializing a `Dictionary`.
- I'm also going to try to build it on iOS using a `DEBUG` build to see if that works.

If you have any more ideas, I'm all ears.

**Comment by ColinSullivan1** [comment]
Would you mind checking if older versions of the .NET client work in your environment - 0.9.0 or 0.8.1?

**Comment by Britvich** [comment]
Unity released 2019.3.0f6 yesterday, so I'm now using that. I'm also on nats-server 2.1.2...

I tried turning code stripping off and that didn't help. I didn't even get the `Disconnected` event from NATS.

I just tried it with NATS .NET client 0.9.0 and get this on the initial open:

```
[74602] 2020/01/28 23:55:42.916365 [DBG] 66.60.161.134:47271 - cid:7 - Client connection created
[74602] 2020/01/28 23:55:43.066480 [DBG] 66.60.161.134:47271 - cid:7 - Client connection closed
```

```
0020/01/28 15:55:38.712 31214 31235 Info Unity SystemInfo CPU = ARM64 FP ASIMD AES, Cores = 8, Memory = 3662mb
2020/01/28 15:55:38.712 31214 31235 Info Unity SystemInfo ARM big.LITTLE configuration: 4 big (mask: 0xf0), 4 little (mask: 0xf)
2020/01/28 15:55:38.712 31214 31235 Info Unity Built from '2019.3/staging' branch, Version '2019.3.0f6 (27ab2135bccf)', Build type 'Release', Scripting Backend 'il2cpp', CPU 'arm64-v8a', Stripping 'Enabled'
2020/01/28 15:55:42.208 31214 31235 Error Unity Unable to find advapi32
2020/01/28 15:55:42.211 31214 31235 Warn Unity Plugins: Couldn't open advapi32, error: dlopen failed: library "advapi32" not found
2020/01/28 15:55:42.211 31214 31235 Warn Unity Microsoft.Win32.ManifestEtw:EventRegister(Guid&, EtwEnableCallback, Void*, Int64&)
2020/01/28 15:55:42.211 31214 31235 Warn Unity System.Diagnostics.Tracing.EventProvider:Register(Guid)
2020/01/28 15:55:42.211 31214 31235 Warn Unity System.Diagnostics.Tracing.EventSource:Initialize(Guid, String, String[])
2020/01/28 15:55:42.211 31214 31235 Warn Unity System.Collections.Concurrent.CDSCollectionETWBCLProvider:.cctor()
2020/01/28 15:55:42.211 31214 31235 Warn Unity System.Collections.Concurrent.ConcurrentDictionary`2:AcquireAllLocks(Int32&)
2020/01/28 15:55:42.211 31214 31235 Warn Unity System.Collections.Concurrent.ConcurrentDictionary`2:GetValues()
2020/01/28 15:55:42.211 31214 31235 Warn Unity NATS.Client.Connection:close(ConnState, Boolean)
2020/01/28 15:55:42.211 31214 31235 Warn Unity NATS.Client.Connection:connect(Srv, Exception&)
2020/01/28 15:55:42.211 31214 31235 Warn Unity System.Predicate`1:Invoke(T)
2020/01/28 15:55:42.211 31214 31235 Warn Unity NATS.Client.ServerPool:ConnectToAServer(Predicate`1)
2020/01/28 15:55:42.211 31214 31235 Warn Unity NATS.Client.Connection:connect()
2020/01/28 15:55:42.211 31214 31235 Warn Unity NATS.Client.ConnectionFactory:CreateConnection(Options)
2020/01/28 15:55:42.211 31214 31235 Warn Unity Worlds.Services:.ctor()
2020/01/28 15:55:42.211 31214 31235 Warn Unity Director:Awake()
2020/01/28 15:55:42.211 31214 31235 Warn Unity UnityEngine.GameObject:AddComponent()
2020/01/28 15:55:42.211 31214 31235 Warn Unity Director:Initialize()
2020/01/28 15:55:42.211 31214 31235 Warn Unity 
2020/01/28 15:55:42.211 31214 31235 Warn Unity [ line 306] 
2020/01/28 15:55:42.211 31214 31235 Warn Unity (Filename:  Line: 306)
2020/01/28 15:55:42.211 31214 31235 Warn Unity 
2020/01/28 15:55:42.237 31214 31235 Error Unity NullReferenceException: Object reference not set to an instance of an object.
2020/01/28 15:55:42.237 31214 31235 Error Unity   at System.Runtime.Serialization.Json.JsonFormatWriterInterpreter.TryWritePrimitive (System.Type type, System.Func`1[TResult] value, System.Reflection.MemberInfo memberInfo, System.Nullable`1[T] arrayItemIndex, System.Xml.XmlDictionaryString name, System.Int32 nameIndex) [0x00000] in <00000000000000000000000000000000>:0 
2020/01/28 15:55:42.237 31214 31235 Error Unity   at System.Runtime.Serialization.Json.JsonFormatWriterInterpreter.WriteMembers (System.Runtime.Serialization.ClassDataContract classContract, System.Runtime.Serialization.ExtensionDataObject extensionData, System.Runtime.Serialization.ClassDataContract derivedMostClassContract) [0x00000] in <00000000000000000000000000000000>:0 
2020/01/28 15:55:42.237 31214 31235 Error Unity   at System.Runtime.Serialization.Json.JsonFormatWriterInterpreter.WriteClass (System.Runtime.Serialization.ClassDataContract classContract) [0x00000] in <00000000000000000000000000000000>:0 
2020/01/28 15:55:42.237 31214 31235 Error Unity   at System.Runtime.Serialization.Json.JsonFormatClassWriterDelegate.Invoke (System.Runtime.Serial
2020/01/28 15:55:42.301 31214 31235 Info Unity ***** Sending login...
2020/01/28 15:55:42.301 31214 31235 Info Unity Startup:.ctor()
2020/01/28 15:55:42.301 31214 31235 Info Unity System.Collections.Generic.KeyCollection:System.Collections.ICollection.CopyTo(Array, Int32)
2020/01/28 15:55:42.301 31214 31235 Info Unity Startup:Start()
2020/01/28 15:55:42.301 31214 31235 Info Unity  
2020/01/28 15:55:42.301 31214 31235 Info Unity (Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)
2020/01/28 15:55:42.301 31214 31235 Info Unity 
2020/01/28 15:55:42.311 31214 31235 Error Unity NullReferenceException: Object reference not set to an instance of an object.
2020/01/28 15:55:42.311 31214 31235 Error Unity   at Startup..ctor () [0x00000] in <00000000000000000000000000000000>:0 
2020/01/28 15:55:42.311 31214 31235 Error Unity   at System.Collections.Generic.Dictionary`2+KeyCollection[TKey,TValue].System.Collections.ICollection.CopyTo (System.Array array, System.Int32 index) [0x00000] in <00000000000000000000000000000000>:0 
2020/01/28 15:55:42.311 31214 31235 Error Unity   at Startup.Start () [0x00000] in <00000000000000000000000000000000>:0 
2020/01/28 15:55:42.311 31214 31235 Error Unity --- End of stack trace from previous location where exception was thrown ---
```

I'm about to try the 0.8.1 client...

**Comment by ColinSullivan1** [comment]
Thanks @Britvich , this looks to be a bug with unity.  Earlier versions of NATS might work (IIRC they don't use a concurrent collection).

**Comment by Britvich** [comment]
I tried it 0.8.1 and got similar results. I've gone back to 0.10.1-pre2. It looks like on my second call to `RequestAsync` it just never returns, and generates an exception later within Unity's `ConcurrentDictionary` code.

Perhaps Unity's `ConcurrentDictionary` implementation is broken. I found these:

https://github.com/grpc/grpc/issues/20553
https://forum.unity.com/threads/il2cpp-build-high-crash-rate-on-il2cpp-os-fastmutex-lock.755597/

I'm wondering how I can work around this. It works fine for debug builds, but the release builds don't work (I've tried using the .NET 4.x and .NET Standard 2.0 API compatibility levels).

Any ideas for us Unity developers that want to use NATS?

**Comment by ColinSullivan1** [comment]
> Perhaps Unity's ConcurrentDictionary implementation is broken. I found these:

I'd put money on that.  I can try to setup a unity env to test, but it may take awhile (I've never used unity tbh).   If you are setup, would you mind writing a code snippet that calls ConcurrentDictionary directly?  One of us can open a unity support ticket and reference the two issues opened up against NATS.net here.

Example code might be as simple as [this](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=netframework-4.8#examples).

**Comment by ColinSullivan1** [comment]
@Britvich, It'd be great to sync up on this issue - I sent you an email.

**Comment by Britvich** [comment]
@ColinSullivan1 Just noticed it in my spam folder. I sent you an email reply and left a voicemail too. Yes, let's get together. I'm making good progress on what's happening.

**Comment by Britvich** [comment]
With the help of @ColinSullivan1 we were able to determine that Unity has a problem with their implementation of `ConcurrentDictionary` on IL2CPP builds. Since NATS uses `ConcurrentDictionary`, Unity IL2CPP builds with NATS won't work.

I have opened a ticket with Unity at:

https://fogbugz.unity3d.com/default.asp?1216719_re63en8l0g4mhcsp

Even this simple example crashes Unity:

```cs
using System.Collections.Concurrent;
using UnityEngine;

public class Startup : MonoBehaviour {
	void Start() {
		var x = new ConcurrentDictionary<int, int>().Values;
	}
}
```

**Workaround:** If you build with the mono libraries, it works fine. If you need to build an IL2CPP version, changing the `ConcurrentDictionary` to a `Dictionary` based version of `Conn.cs` works:

Line 132 becomes:
```cs
private Dictionary<Int64, Subscription> subs = new Dictionary<Int64, Subscription>();
```

Line 3063 becomes:
```cs
if (!subs.ContainsKey(s.sid))
	return;
o = subs[s.sid];
subs.Remove(s.sid);
```

Line 3578 becomes:
```cs
public int SubscriptionCount {
	get {
		lock (mu) {
			return subs.Count;
		}
	}
}
```

Line 3646 adds:
```cs
lock (mu) {
```

Line 3673 adds:
```cs
}
```

**Comment by ColinSullivan1** [comment]
@Britvich , thanks for taking the time to help debug and open the issue.  Much appreciated.  If Unity doesn't move quickly on this issue we can look into working around it but this seems to be a significant bug on their end that should be straightforward to fix.

**Comment by ColinSullivan1** [comment]
@Britvich , I couldn't publicly comment on the issue you filed, but thank you for updating your comment here.  Feel free to link back to this issue on the Unity issue site.  This is a blocking issue for NATS + Unity users, and I'd imagine there are many others that use `ConcurrentDictionary.Values`.

**Comment by ColinSullivan1** [comment]
Unity has been able to reproduce the issue and is looking into it.

https://fogbugz.unity3d.com/default.asp?1216719_re63en8l0g4mhcsp

**Comment by Britvich** [comment]
For those wanting to upvote this problem, it's now also on Unity's issue tracker at:

https://issuetracker.unity3d.com/issues/anroid-logcat-throws-advapi32-related-warning-when-using-concurrentdictionary-in-il2cpp-builds

**Comment by ColinSullivan1** [comment]
I noticed in the unity issue tracker notes that this issue is not reproducible with unity versions 2017.4.36f1, 2018.4.17f1, 2019.3.0b1.

**Comment by ColinSullivan1** [comment]
An update - I received this message from the Unity Customer QA team:

```text
Hello,

I want to inform you that we are unable to reproduce this issue after multiple retests, please update to the latest 2019.3 versions and the fix may be fixed for you. This case will be closed.

Best regards,
Augustinas
Customer QA Team
```

**Comment by Britvich** [comment]
I am able to reproduce this problem on all the latest versions of 2019.3 (including 2019.3.2f1, 2019.3.1f1, 2019.3.0f5). I am using a Google Pixel 2 XL (version 10.0, abi: arm64-v8a, sdk: 29). Here's the logcat output...

![image](https://user-images.githubusercontent.com/7193012/75070735-b2d89a00-54a8-11ea-86ae-802d2dbc00b7.png)

I've let Unity know, and they have reactivated this problem.

**Comment by Britvich** [comment]
@ColinSullivan1 I tried using #367 and now it's erroring while instantiating the `ConcurrentBag`:

<img width="998" alt="Screen Shot 2020-02-27 at 6 04 39 PM" src="https://user-images.githubusercontent.com/7193012/75503870-c04fc180-598b-11ea-9d6e-c3fffff7d693.png">

Good ole Unity... ;)

**Comment by ColinSullivan1** [comment]
Wow, I'm sorry to hear that.  At this point I think Unity needs to take a look at their implementation of Concurrent classes.  They've been available for a long time - since .NET framework 4.0 and .NET core 1.0.

**Comment by Britvich** [comment]
**Update 3**: I am able to run NATS with Unity IL2CPP builds. Here's how:

1. Use the prerelease of 0.10.1 (release 2)
2. Make sure the `UseOldRequestStyle` option is `true`.

It turns out that even though Unity warns about using `ConcurrentDictionary` and `ConcurrentBag`, **they still function correctly**. 

The new request style that nats.net uses **does not work** with Unity. I haven't spent a lot of time figuring out why, but I do know that Unity requires async results to be returned on the main (UI) thread's `SynchronizationContext`. Here's some potentially useful discussions...

https://forum.unity.com/threads/async-await-not-joining-main-thread-after-await.481212/
https://forum.unity.com/threads/using-async-await-with-task-i-created-using-taskcompletionsource.809487/
https://www.reddit.com/r/csharp/comments/9kdkwi/the_danger_of_taskcompletionsourcet_class/
http://www.stevevermeulen.com/index.php/2017/09/using-async-await-in-unity3d-2017/

...and a good video...

https://www.youtube.com/watch?v=J0mcYVxJEl0

With some good comments...

I felt there were some good points but also some misconceptions in this talk.

Excellent points:
1) await causes the method to return.
2) async catches exceptions and places them on the returned task.
3) Fire-and-forget also forgets exceptions.
4) Use GetAwaiter().GetResult() instead of Wait()/Result to avoid the AggregateException.
5) One pitfall of eliding async and await is try/catch.
6) ValueTask is good for synchronous hot paths.
7) The guideline is "avoid async void", not "never use async void".
8) Some form of IAsyncCommand/AsyncCommand is useful for MVVM apps.

There were also a couple of misconceptions. One big one is that another thread is used to block for asynchronous code; this is usually not the case. Another is the role that SynchronizationContext plays. Detailed points:

1) SynchronizationContext does not manage the thread pool. It is a conceptual "context" that may represent a UI context or a thread pool context or any other context. https://msdn.microsoft.com/en-us/magazine/gg598924.aspx
2) While the thread pool does have a maximum number of threads, this is almost never a problem. Usually, work queues up due to the limited thread injection rate rather than hitting the maximum number of threads.
3) @5:00, DownloadDataTaskAsync doesn't require a thread blocking on the download. There is no "Thread 2". https://blog.stephencleary.com/2013/11/there-is-no-thread.html
4) @7:45, A SynchronizationContext does not necessarily resume on the original thread. This example is only true for UI threads, and I think it would be clearer to say it's a UI SynchronizationContext rather than "the SynchronizationContext" asking "what thread called you?" and attempting to resume the method back on the original thread. https://blog.stephencleary.com/2012/02/async-and-await.html
5) @18:20 and @19:40, async methods do not run on other threads. Using await is correct, but not because it uses fewer threads. https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html
6) @22:20 ConfigureAwait(false) avoids the SynchronizationContext *completely*. The SynchronizationContext is not involved at all. ConfigureAwait(true) does not always return to the calling thread; that is only true for a UI context.
7) @24:50 Eliding async and await does not save a context switch, since there are no other threads involved here. It does provide a minor performance benefit by avoiding the state machine. However, eliding async/await has some other pitfalls not mentioned in this talk, in particular when exceptions are thrown and how they can be caught: https://blog.stephencleary.com/2016/12/eliding-async-await.html
8) ValueTask has pitfalls as well, some of which can be surprising to devs used to task: https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/
9) @34:50 There aren't two threads accessing TopStoryList; there's no "Thread 5" and no race condition. It's always the same UI thread, though the order of access can be surprising and cause bugs.

In summary, I felt this was a pretty good talk and provided some good guidelines. The reasoning behind some of the guidelines isn't quite right, but as long as listeners avoid the misconceptions mentioned above, the guidelines themselves will serve them well.

**Comment by Britvich** [comment]
**Update 4:** It still fails sometimes even when the `UseOldRequestStyle` option is `true`.

I believe that after `await`ing on some requests, the results are returned on a different thread than Unity's UI `SynchronizationContext` main thread. All bets are off when that happens. I will work on this more in the morning... too tired.

**Comment by ColinSullivan1** [comment]
@Britvich  - Interesting - thank you for the details here, very much appreciated!  We could look at changing the old request style to avoid the await operator entirely - the original code did not use await.  

imo, this is a serious shortcoming of Unity.

**Comment by Britvich** [comment]
@ColinSullivan1 I don't recommend avoiding `async` `awaits`.

The root of the problem is that some .NET frameworks (such as ASP.NET, WPF, Windows Forms, Windows RunTime [WinRT], xunit, and Unity), require you to honor their synchronization model via the `SynchronizationContext` class. A .NET `Task` automatically honors whatever `SynchronizationContext` exists (sometimes it's null). But NATS overrides the context through calls like: `ConfigureAwait(false)`, and therefore when NATS returns from an await, it's often on the wrong thread.

This should be a pretty easy fix. To honor the `SynchronizationContext`, if NATS used the `SynchronizationContext.Post()` method (that these frameworks override), everything will work correctly. See:

http://hamidmosalla.com/2018/06/24/what-is-synchronizationcontext/
https://devblogs.microsoft.com/dotnet/configureawait-faq/

Until that is honored, all .NET frameworks (not just Unity) that require a synchronization model will be unable to use NATS.

**Comment by Britvich** [comment]
@ColinSullivan1 Could we meet online again? I've been trying to get this solved and am very close. I'd like to show you what's happening and get your thoughts...

**Comment by ColinSullivan1** [comment]
Absolutely - Thanks!  I'd like to knock this out.  Are you free tomorrow AM?  Feel free to send email with a good time and I can setup a google hangouts to screen share.

**Comment by Britvich** [comment]
This has been fixed. See https://github.com/nats-io/nats.net/pull/370

**Comment by ColinSullivan1** [comment]
Thank you so much for the contribution and the many hours spent tracking this down.  Much appreciated!

**Comment by TigerHix** [comment]
We are having very similar error logs on Unity IL2CPP in our own project. Is the issue related to the use of ConcurrentDictionary? Or the `Could not find advapi32` is just a harmless warning, as suggested by the Unity team in their [issue](https://issuetracker.unity3d.com/issues/desktop-log-of-warn-unity-plugins-couldnt-open-advapi32-error-dlopen-failed-library-advapi32-not-found)?

**Comment by Britvich** [comment]
NATS took out their use of `ConcurrentDictionary`. I think Unity fixed things so `ConcurrentDictionary` works fine in their latest versions (2020.1.x and 2020.2.x).
